File: web/src/api/knowledge.ts
// 知识库 RAG API 封装（M5-02）：知识库 CRUD + 文档索引/检索。
// 后端契约见 server/internal/api/knowledge.go。
import { request } from './client'

export interface KnowledgeBase {
  id: number
  user_id: number
  name: string
  description: string
  embedding_model: string
  doc_count: number
  chunk_count: number
  created_at: string
  updated_at: string
}

export interface DocInfo {
  name: string
  chunk_count: number
}

export interface SearchHit {
  content: string
  score: number
  source: string
  chunk_index: number
}

export async function listKnowledgeBases(): Promise<KnowledgeBase[]> {
  const data = await request<{ knowledge_bases: KnowledgeBase[]; total: number }>('/knowledge')
  return data.knowledge_bases ?? []
}

export async function createKnowledgeBase(name: string, description: string): Promise<KnowledgeBase> {
  return request<KnowledgeBase>('/knowledge', { method: 'POST', body: { name, description } })
}

export async function updateKnowledgeBase(
  id: number,
  name: string,
  description: string,
): Promise<KnowledgeBase> {
  return request<KnowledgeBase>(`/knowledge/${id}`, {
    method: 'PUT',
    body: { name, description },
  })
}

export async function deleteKnowledgeBase(id: number): Promise<void> {
  await request(`/knowledge/${id}`, { method: 'DELETE' })
}

export async function listDocuments(id: number): Promise<DocInfo[]> {
  const data = await request<{ documents: DocInfo[]; total: number }>(`/knowledge/${id}/documents`)
  return data.documents ?? []
}

export async function indexDocument(
  id: number,
  name: string,
  content: string,
  contentType: string,
): Promise<{ indexed_chunks: number }> {
  return request<{ indexed_chunks: number }>(`/knowledge/${id}/documents`, {
    method: 'POST',
    body: { name, content, content_type: contentType },
  })
}

export async function deleteDocument(id: number, name: string): Promise<{ deleted_chunks: number }> {
  return request<{ deleted_chunks: number }>(
    `/knowledge/${id}/documents/${encodeURIComponent(name)}`,
    { method: 'DELETE' },
  )
}

export async function searchKnowledge(
  id: number,
  query: string,
  topK: number,
): Promise<SearchHit[]> {
  const data = await request<{ hits: SearchHit[]; total: number }>(`/knowledge/${id}/search`, {
    method: 'POST',
    body: { query, top_k: topK },
  })
  return data.hits ?? []
}

File: server/internal/model/knowledge.go
package model

import (
	"errors"
	"time"

	"gorm.io/gorm"
)

// KnowledgeBase 是用户的私有知识库（M5-02 Knowledge RAG）。
// 一个知识库可索引多份文档（按 chunk 切片后存入向量库），对话时按用户归属检索注入。
type KnowledgeBase struct {
	ID             uint           `gorm:"primaryKey" json:"id"`
	UserID         uint           `gorm:"index;not null" json:"user_id"` // 归属用户（owner 隔离）
	Name           string         `gorm:"size:191;not null" json:"name"`
	Description    string         `gorm:"size:512" json:"description"`
	EmbeddingModel string         `gorm:"size:64;not null;default:local-hashing" json:"embedding_model"` // 嵌入模型标识（本地基线 local-hashing，预留 pgvector/远程升级位）
	DocCount       int            `gorm:"not null;default:0" json:"doc_count"`                          // 文档（源）数
	ChunkCount     int            `gorm:"not null;default:0" json:"chunk_count"`                        // 切片（向量）数
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName 固定表名，避免 GORM 复数化规则变化影响既有库。
func (KnowledgeBase) TableName() string { return "knowledge_bases" }

// Validate 校验知识库创建/更新参数的合法性。
func (kb *KnowledgeBase) Validate() error {
	if kb.Name == "" {
		return errors.New("知识库名称不能为空")
	}
	if len(kb.Name) > 191 {
		return errors.New("知识库名称过长")
	}
	if len(kb.Description) > 512 {
		return errors.New("知识库描述过长")
	}
	if kb.EmbeddingModel == "" {
		kb.EmbeddingModel = "local-hashing"
	}
	return nil
}
